[Docs] Add advance doc for model (#415)
open-mmlab/mmgeneration: docs/en/advanced_guides/models.md

# Design Your Own Models

MMGeneration is built upon MMEngine and MMCV, which enables users to design new models quickly, train and evaluate them easily. In this section, you will learn how to design your own models.

## Overview of models in MMGeneration

> In MMGeneration, one algorithm can be splited two compents: **Model** and **Architecture**(s).

Reviewer: remove (s)

Reviewer: Model and Module

- **Model** are topmost wrappers and always inherint from `BaseModel` provided in MMEngine. **Model** is responsible to network forward, loss calculation and backward, parameters updating, etc. In MMGeneration, **Model** should be registered as `MODELS`.
- **Architecture**(s) is the neural network to train or inference, and always as element(s) of **Model**. In MMGeneration, **Architecture** should be registered as **MODULES**.

> Take BigGAN model as an example, [generator and discriminator](https://github.com/open-mmlab/mmgeneration/blob/test-1.x/mmgen/models/architectures/biggan/generator_discriminator.py), [`BigGAN`](https://github.com/open-mmlab/mmgeneration/blob/test-1.x/mmgen/models/gans/biggan.py) is the **Model**, which take data from dataloader and train generator and discriminator alternatively.

Reviewer: ```suggestion
Take BigGAN model as an example, [generator and discriminator](https://github.com/open-mmlab/mmgeneration/blob/test-1.x/mmgen/models/architectures/biggan/generator_discriminator.py) is the **Module**, which generate images and discriminate real or fake images. [`BigGAN`](https://github.com/open-mmlab/mmgeneration/blob/test-1.x/mmgen/models/gans/biggan.py) is the **Model**, which take data from dataloader and train generator and discriminator alternatively.
```

> Take BigGAN model as an example, [generator and discriminator](https://github.com/open-mmlab/mmgeneration/blob/test-1.x/mmgen/models/architectures/biggan/generator_discriminator.py), [`BigGAN`](https://github.com/open-mmlab/mmgeneration/blob/test-1.x/mmgen/models/gans/biggan.py) is the **Model**, which take data from dataloader and train generator and discriminator alternatively.

Reviewer: You can just take dcgan as an example.

> You can find implementation **Model** for [GAN](https://github.com/open-mmlab/mmgeneration/tree/test-1.x/mmgen/models/gans) and [Diffusion](https://github.com/open-mmlab/mmgeneration/tree/test-1.x/mmgen/models/architectures) models here. And implementation of **Architecture**(s) can be find [here](https://github.com/open-mmlab/mmgeneration/tree/test-1.x/mmgen/models/architectures).

Reviewer: ```suggestion
You can find implementation **Model** for [GAN](https://github.com/open-mmlab/mmgeneration/tree/dev-1.x/mmgen/models/gans) and [Diffusion](https://github.com/open-mmlab/mmgeneration/tree/dev-1.x/mmgen/models/diffusion) models here. And implementation of **Architecture**(s) can be find [here](https://github.com/open-mmlab/mmgeneration/tree/dev-1.x/mmgen/models/architectures).
```

Reviewer: Actually we offer ``gans``, ``translation_models``, ``diffusions``, ``losses``, ``architectures`` , these should all be covered.

Reviewer: You can directly give a list like these.
Model
   xxx()
Module
   xxx()
Denote we should mention that losses are regarded as MODULES im MMGeneration.

> Here, we take the implementation of the classical gan model, DCGAN \[1\], as an example.

Reviewer: Above docs say you take biggan as example. here should be the same.

        elif pretrained is None:
            for m in self.modules():
                if isinstance(m, (nn.Conv2d, nn.ConvTranspose2d)):
                    normal_init(m, 0, 0.02)
                elif isinstance(m, _BatchNorm):
                    nn.init.normal_(m.weight.data)
                    nn.init.constant_(m.bias.data, 0)
        else:
            raise TypeError('pretrained must be a str or None but'
                            f' got {type(pretrained)} instead.')
```

After the implementation of class `DCGANGenerator`, we need to update the model list in `mmgen/models/architectures/__init__.py`, so that we can import and use class `DCGANGenerator` by `mmgen.models.architectures`.

Implementation of Class `DCGANDiscriminator` follows the similar logic, and you can find the implementation [here](https://github.com/open-mmlab/mmgeneration/blob/test-1.x/mmgen/models/architectures/dcgan/generator_discriminator.py#L195).

## Step 2: Define the your **Model**

After the implementation of the network **Architecture**, we need to define our **Model** class `DCGAN`.

In **Model**, you should implement three function `train_step`, `val_step` and `test_step`.

> - `train_step`: This function is responsible to update the parameters of the network and called by `IterBasedTrainLoop` or `EpochBasedTrainLoop`. `train_step` take data batch as input and return a dict of log.

Reviewer: mmengine's Loop

> - `val_step`: This function is responsible for getting output for validation during the training process. and is called by `GenValLoop`.

Reviewer: GenValLoop is not known to users, we may attach a link to docs or codes.

> - `test_step`: This function is responsible for getting output in test process and is called by `GenTestLoop`.

Reviewer: `GenTestLoop` is not known to users, we may attach a link to docs or codes.

> For simplify using, we provide `BaseGAN` class im MMGeneration, which implement generic `train_step`, `val_step` and `test_step` function for GAN models. With `BaseGAN`, each specific GAN algorithm only need to implement `train_generator` and `train_discriminator`.

Reviewer: ```suggestion
For simple usage, we provide `BaseGAN` class im MMGeneration, which implements generic `train_step`, `val_step` and `test_step` function for GAN models. With `BaseGAN` as base class, each specific GAN algorithm only need to implement `train_generator` and `train_discriminator`.
```

> In `train_step`, we support data preprocessing, gradient accumulation (realized by [`OptimWrapper`](https://github.com/open-mmlab/mmengine/blob/main/docs/en/tutorials/optim_wrapper.md)) and expontial moving averate (EMA) realized by [(`ExponentialMovingAverage`)](https://github.com/open-mmlab/mmgeneration/blob/test-1.x/mmgen/models/averaged_model.py#L19). With `BaseGAN.train_step`, each specific GAN algorithm only need to implement `train_generator` and `train_discriminator`.

Reviewer: ```suggestion
In `train_step`, we support data preprocessing, gradient accumulation (realized by [`OptimWrapper`](https://github.com/open-mmlab/mmengine/blob/main/docs/en/tutorials/optim_wrapper.md)) and expontial moving averate (EMA) realized by [(`ExponentialMovingAverage`)](https://github.com/open-mmlab/mmgeneration/blob/dev-1.x/mmgen/models/averaged_model.py#L19). With `BaseGAN.train_step`, each specific GAN algorithm only need to implement `train_generator` and `train_discriminator`.
```

Reviewer: seems we have no introduction for data preprocessing in whole documents, we may give its introduction here.
